boards/shared/Src/Io/Io_SharedFreqOnlyPwmInput.c

```diff
@@ -0,0 +1,132 @@
+    uint32_t           rising_edge_timer_channel,
+    uint32_t           timer_auto_reload_reg)
+{
+    assert(htim != NULL);
+    static struct FreqOnlyPwmInput pwm_inputs[MAX_NUM_OF_FREQ_ONLY_PWM_INPUTS];
+    static size_t                  alloc_index = 0;
+
+    assert(alloc_index < MAX_NUM_OF_FREQ_ONLY_PWM_INPUTS);
+
+    struct FreqOnlyPwmInput *const pwm_input = &pwm_inputs[alloc_index++];
+
+    pwm_input->frequency_hz            = NAN;
+    pwm_input->htim                    = htim;
+    pwm_input->timer_frequency_hz      = timer_frequency_hz;
+    pwm_input->rising_edge_tim_channel = rising_edge_timer_channel;
+    pwm_input->timer_auto_reload_reg   = timer_auto_reload_reg;
+    pwm_input->prev_rising_edge        = 0U;
+    pwm_input->curr_rising_edge        = 0U;
+    pwm_input->first_tick              = true;
+    pwm_input->elapsed_tick            = 0U;
+
+    HAL_TIM_IC_Start_IT(htim, rising_edge_timer_channel);
+    HAL_TIM_Base_Start_IT(htim);
+
+    return pwm_input;
+}
+
+void Io_SharedFreqOnlyPwmInput_Tick(struct FreqOnlyPwmInput *const pwm_input)
+{
+    // Reset elapsed tick counter for the flow meter
+    pwm_input->elapsed_tick = 0U;
+
+    if (pwm_input->first_tick)
+    {
+        pwm_input->curr_rising_edge = HAL_TIM_ReadCapturedValue(
+            pwm_input->htim, pwm_input->rising_edge_tim_channel);
+        pwm_input->first_tick = false;
+    }
+    else
+    {
+        pwm_input->prev_rising_edge = pwm_input->curr_rising_edge;
+        pwm_input->curr_rising_edge = HAL_TIM_ReadCapturedValue(
+            pwm_input->htim, pwm_input->rising_edge_tim_channel);
+
+        uint32_t       rising_edge_delta;
+        const uint32_t prev_rising_edge = pwm_input->prev_rising_edge;
+        const uint32_t curr_rising_edge = pwm_input->curr_rising_edge;
+
+        // Check if the counter value has overflowed
+        if (curr_rising_edge > prev_rising_edge)
+        {
+            rising_edge_delta = curr_rising_edge - prev_rising_edge;
+            Io_SetFrequency(
+                pwm_input,
+                (float)pwm_input->timer_frequency_hz / rising_edge_delta);
+        }
+        else if (curr_rising_edge < prev_rising_edge)
+        {
+            rising_edge_delta = pwm_input->timer_auto_reload_reg -
+                                prev_rising_edge + curr_rising_edge + 1;
+            Io_SetFrequency(
+                pwm_input,
+                (float)pwm_input->timer_frequency_hz / rising_edge_delta);
+        }
+        else
+        {
+            //Occurs when the counter rolls over, or either when a tick arrives before the counter can upcount
+            Io_SetFrequency(pwm_input, NAN);
+        }
+    }
+}
+
+void Io_SharedFreqOnlyPwmInput_Elapsed_Tick(
+    struct FreqOnlyPwmInput *const pwm_input)
+{
+    // Elapsed tick counts the number of times where the timer counter overflows
+    const size_t elapsed_tick = ++pwm_input->elapsed_tick;
+
+    if (elapsed_tick >= MAX_TIMER_PERIOD_ELAPSED_COUNT)
```

Comment:
```suggestion
    // If the timer overflows twice without a rising edge, the PWM signal is likely inactive and its frequency can't be computed
    if (elapsed_tick >= 2U)
```
Don't even really need a macro `MAX_TIMER_PERIOD_ELAPSED_COUNT` here. What would help with readability is describing what the value of 2 represents.